lora cuda multistream

hai-meh-cs opened the pull request:
## Purpose

 This PR introduces an optional dedicated CUDA stream for Punica LoRA computation and wires it into the LoRA linear path so that base and LoRA GEMMs can overlap.
FIX [#27371](https://github.com/vllm-project/vllm/issues/27371)
  Concretely:

  - `PunicaWrapperGPU` gets:
    - A dedicated `lora_stream` (using `aux_stream()`), gated by a new env var `VLLM_DISABLE_LORA_STREAM`.
    - Helpers `_run_on_lora_stream(fn)` and `_run_on_lora_stream_async(fn, pre_event)` that:
      - Run Punica kernels on the LoRA stream and synchronize with the default stream only when needed.
      - Fall back to the current stream when `torch.compile` / Dynamo is tracing, or when the LoRA stream is disabled.
    - A small NVTX helper `_nvtx(name)` controlled by `VLLM_LORA_NVTX`:
      - Emits NVTX ranges around LoRA launches for Nsight Systems / Nsight Compute.
      - Becomes a no-op when `torch._dynamo.is_compiling()` or `torch.compiler.is_compiling()` is true to avoid graph breaks from `range_push / range_pop`.

  - `BaseLinearLayerWithLoRA.apply` is updated to overlap base and LoRA computation:
    - Records a `pre_event` on the current stream before launching the base GEMM.
    - Runs the base GEMM on the default stream as today.
    - Launches LoRA shrink/expand on the LoRA stream into a separate buffer via a new `add_lora_linear_into_async(...)` API on `PunicaWrapperGPU`, passing `pre_event` so LoRA computation sees its inputs without waiting on the base GEMM.
    - After the async LoRA event completes, performs `output.add_(lora_delta)` to combine base and LoRA.
    - Falls back to the original synchronous path when the async helper is not available.

  - Semantics are intentionally unchanged:
    - We still compute `W x + ΔW x` for all LoRA-wrapped linears.
    - The change is purely in scheduling (overlap and NVTX instrumentation), not in numerical output.

  - A new env var `VLLM_DISABLE_LORA_STREAM` is added in `envs.py` so operators can easily turn the LoRA stream off and go back to single-stream behavior for debugging / benchmarking

 ## Test Plan

 Environment:

  - Remote setup running vLLM 0.11.2 (Python 3.12, CUDA, `cudagraph_mode=FULL_AND_PIECEWISE`) with:
    - Base model: `dphn/Dolphin3.0-Llama3.2-1B`
    - LoRA: `AdityaKothari/Llama-3.2-1B-Sentiment-Analysis`
    - Punica LoRA enabled (`--enable-lora`, `--max-loras`, `--max-lora-rank`, etc.)

  Functional / integration:

  - Start vLLM with LoRA enabled and CUDA graphs turned on:

    ```bash
    VLLM_LORA_NVTX=0 \
    python3 -m vllm.entrypoints.openai.api_server \
      --model dphn/Dolphin3.0-Llama3.2-1B \
      --tensor-parallel-size 1 \
      --max-model-len 4096 \
      --enable-lora --max-loras 4 --max-lora-rank 64 \
      --lora-extra-vocab-size 256 --max-cpu-loras 4 \
      --lora-modules nemo20b=AdityaKothari/Llama-3.2-1B-Sentiment-Analysis \
      --no-enable-prefix-caching

  - Verify that the engine initializes successfully with cudagraph_mode enabled and Punica LoRA (PunicaWrapperGPU) selected.
  - Run basic chat/completion requests against /v1/chat/completions to confirm correct behavior with LoRA enabled.
  - Run inference-perf matrices to ensure stability under load (no crashes, consistent latencies/throughput):

    ### On a separate host using the OpenAI-compatible endpoint:
    DURATION=30 OUTPUT_TOKENS=512 RATE_FACTOR=4 \
      ./scripts/run_inference_perf_matrix.sh
      - This sweeps prompt lengths 128 / 256 / 1024 and concurrency 1 / 2 / 4 / 8.
      - Confirm that:
          - All runs complete successfully and produce summary_lifecycle_metrics.json.
          - Latency/throughput curves are consistent with the single-stream baseline (no obvious regressions for common workloads).

  Profiling (optional, to validate overlap and NVTX):

  - Start vLLM with NVTX enabled:

    VLLM_LORA_NVTX=1 \
    python3 -m vllm.entrypoints.openai.api_server \
      ...same arguments as above...
  - Profile a short run with Nsight Systems:

    VLLM_LORA_NVTX=1 \
    nsys profile -t cuda,nvtx -o nsys_lora \
      python3 -m vllm.entrypoints.openai.api_server \
        ...same arguments...
  - In Nsight Systems:
      - Confirm the presence of LoRA ranges (e.g., lora_launch_async / lora_launch_sync).
      - Verify that LoRA shrink/expand kernels on the LoRA stream overlap with base GEMMs on the default stream as expected.
      - Verify that during torch.compile / CUDA graph capture there are no NVTX ranges (the helper no-ops while compiling), so there are no Dynamo graph breaks from NVTX.

  Behavior with the LoRA stream disabled:

  - Run with:

    VLLM_DISABLE_LORA_STREAM=1 \
    python3 -m vllm.entrypoints.openai.api_server ...
  - Confirm that LoRA still works and behavior matches the pre-PR single-stream scheduling.

  ## Test Result

  - Engine:
      - Engine and CUDA graph capture initialize successfully on vLLM 0.11.2 with LoRA enabled and cudagraph_specialize_lora=True.
      - LoRA-augmented generations (Punica) behave as expected under both single-stream and multi-stream configurations (qualitative spot checks
        on outputs).
  - Load/Perf:
      - inference-perf matrix runs for Llama3.2 1B + LoRA succeed across:
          - Prompt lengths: 128, 256, 1024
          - Concurrency: 1, 2, 4, 8
      - Request latency and throughput scale appropriately; no functional regressions or stability issues observed.
      - Preliminary numbers show similar or slightly improved throughput at higher concurrencies, with overlap visible in Nsight Systems traces.
  - Profiling:
      - Nsight Systems traces show base GEMMs on the default stream and LoRA shrink/expand on the LoRA stream with NVTX ranges that confirm
        overlap.
      - During torch.compile / CUDA graph capture, _nvtx correctly becomes a no-op and does not introduce Dynamo graph breaks.

  ———
  <details>
  <summary> Essential Elements of an Effective PR Description Checklist </summary>

  - [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
  - [x] The test plan, such as providing test command.
  - [x] The test results, such as pasting the results comparison before and after, or e2e results.
  - [ ] (Optional) The necessary documentation update, such as updating supported_models.md and examples for a new model.
  - [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the Google Doc.
  </details>

heheda12345 commented:
CC @jeejeelee

ProExpertProg reviewed (changes requested):
The diff looks polluted, can you merge/rebase from main?

jaislas-a committed:
feat(lora): Add CUDA multistream parallel execution for LoRA

Implements true parallel execution of LoRA computation alongside base GEMM
using CUDA streams, following the MoE shared_experts pattern.

Key changes:
- Custom op wraps ENTIRE parallel execution (base GEMM + LoRA)
- LoRA computation starts BEFORE base GEMM for true parallelism
- Uses vLLM's aux_stream() for LoRA computation
- Opt-in via VLLM_ENABLE_LORA_STREAM=1 environment variable

Execution timeline:
  Main stream:  [clone x] -> [base GEMM] ---------> [wait] -> [add]
  Aux stream:              -> [LoRA: B @ (A @ x)] --^
                              PARALLEL!

Benchmark results (gpt-oss-20b with GSM8K LoRA rank=32):
- OTPS (Output Tokens/s): +2.9% improvement
- ITL (Inter-Token Latency): +2.8% improvement
- E2E Latency: +1.8% improvement
- Best gains at higher concurrency (conc=4-8)

Note: TTFT may increase at higher concurrency due to stream
synchronization overhead. Recommended for decode-heavy workloads.

jeejeelee commented on vllm/lora/layers/base_linear.py:
```diff
@@ -119,39 +214,110 @@ def set_lora(
 
     @property
     def weight(self) -> torch.Tensor:
-        # unquantizedLinear
```
QQ: why delete these comments?

jeejeelee commented on vllm/lora/layers/base_linear.py:
```diff
@@ -1,45 +1,144 @@
     RowParallelLinear,
 )
 from vllm.platforms import current_platform
+from vllm.utils.torch_utils import aux_stream, direct_register_custom_op
```
Could you please update the usage of `aux_stream`  in https://github.com/vllm-project/vllm/blob/main/vllm/utils/torch_utils.py#L416

jeejeelee commented on vllm/lora/layers/base_linear.py:
```diff
@@ -1,45 +1,144 @@
         self.n_slices: int
 
+        # Use vLLM's global aux_stream for parallel execution
+        self._lora_stream = aux_stream() if _ENABLE_LORA_STREAM else None
```
Since aux_stream was designed for sharded experts, I'm curious about what happens when we enable LoRA for an MoE model with sharded expert

jeejeelee reviewed (commented):
Sorry for the delay feedback, I add some init comments here
